File: backend/app/api/v1/feedback.py
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.db.mongo import get_database
from app.models.feedback import NegotiatedRent, NegotiatedRentIn, UserFeedback, UserFeedbackIn

router = APIRouter()


@router.post("/negotiated-rents", response_model=NegotiatedRent)
async def submit_negotiated_rent(payload: NegotiatedRentIn, db: AsyncIOMotorDatabase = Depends(get_database)) -> NegotiatedRent:
    doc = NegotiatedRent(**payload.model_dump()).model_dump(by_alias=True, exclude_none=True)
    result = await db.negotiated_rents.insert_one(doc)
    doc["_id"] = result.inserted_id
    return NegotiatedRent.model_validate(doc)


@router.post("/locality", response_model=UserFeedback)
async def submit_locality_feedback(payload: UserFeedbackIn, db: AsyncIOMotorDatabase = Depends(get_database)) -> UserFeedback:
    doc = UserFeedback(**payload.model_dump()).model_dump(by_alias=True, exclude_none=True)
    result = await db.user_feedback.insert_one(doc)
    doc["_id"] = result.inserted_id
    return UserFeedback.model_validate(doc)

File: backend/src/api/v1/feedback.ts
import { Router, Request, Response } from "express";
import { getDatabase } from "../../db/mongo.js";
import {
  NegotiatedRentInSchema,
  UserFeedbackInSchema,
} from "../../models/feedback.js";
import { serializeDoc } from "../../repositories/base.js";
import { emailService } from "../../services/emailService.js";

export const feedbackRouter = Router();

// POST /api/v1/feedback/negotiated-rents
feedbackRouter.post("/negotiated-rents", async (req: Request, res: Response): Promise<void> => {
  try {
    const parsed = NegotiatedRentInSchema.parse(req.body);
    const db = getDatabase();
    const doc = {
      ...parsed,
      created_at: new Date().toISOString(),
    };
    const result = await db.collection("negotiated_rents").insertOne(doc);

    // Notify admin email asynchronously
    emailService.sendNegotiatedRentAlert({
      property_id: parsed.property_id,
      original_rent: parsed.original_rent,
      negotiated_rent: parsed.negotiated_rent,
      notes: parsed.notes,
    }).catch(() => {});

    res.json(serializeDoc({ ...doc, _id: result.insertedId }));
  } catch (error) {
    res.status(400).json({ error: (error as Error).message });
  }
});

// POST /api/v1/feedback/locality
feedbackRouter.post("/locality", async (req: Request, res: Response): Promise<void> => {
  try {
    const parsed = UserFeedbackInSchema.parse(req.body);
    const db = getDatabase();
    const doc = {
      ...parsed,
      created_at: new Date().toISOString(),
    };
    const result = await db.collection("user_feedback").insertOne(doc);

    // Notify admin email asynchronously
    emailService.sendLocalityFeedbackAlert({
      locality_id: parsed.locality_id,
      score: parsed.score,
      comment: parsed.comment,
    }).catch(() => {});

    res.json(serializeDoc({ ...doc, _id: result.insertedId }));
  } catch (error) {
    res.status(400).json({ error: (error as Error).message });
  }
});

